File: TypeScript/LLMAgent.ts
// Please install OpenAI SDK first: `npm install openai`
import OpenAI from "openai";
const baseURL = '';
const apiKey = '';
import { Project, SyntaxKind } from "ts-morph";
import { ProjectDataLoader } from "./DataLoader/ProjectDataLoader"
import { ThirdPackageLoader } from "./DataLoader/ThirdPackageLoader"
import { Logger, LogLevel } from "./Util/logMethods"
import { isQuotedUnionBySplit } from "./Util/tools"
const logger = new Logger({
    level: LogLevel.DEBUG,
    format: "{time} [{level}] ▶ {message}",
});
export class LLMAgent {
    private openai: OpenAI;
    private total_prompt: string = "";
    private basePrompt = "Next, you will be provided with a piece of TypeScript code slice. You will infer the variable type or function return type in TypeScript and fill in the type annotation in <mask>. Output just only the type ,which you infer, nothing else,\n"
    private example_prompt = "Example output: mask: string\n"
    private filePath: string = ""
    private projectDataLoader: ProjectDataLoader;
    constructor() {
        this.openai = new OpenAI({ baseURL: baseURL, apiKey: apiKey });
        this.projectDataLoader = new ProjectDataLoader();
    }
    async Generation(prompts: string[], CODE: string) {
        let totalPrompt = this.basePrompt + this.example_prompt
        prompts.forEach(prompt => {
            totalPrompt += `${prompt}\n`
        });
        totalPrompt += `The code you need to make a prediction is:\n${CODE}`
        // const openai = new OpenAI({ baseURL: baseURL, apiKey: apiKey });
        this.total_prompt = totalPrompt
        const completion = await this.openai.chat.completions.create({
            messages: [{ role: "user", content: totalPrompt }],
            model: "gpt-3.5-turbo",
            // model: "deepseek-chat",
            // model: "gpt-4o-mini",
            // model: "claude-3-haiku-20240307",
            // model: "qwen3-coder-plus",
            max_tokens: 50,
            temperature: 0.2,
            top_p: 0.3,
        });
        logger.debug(`total prompt:\n${totalPrompt}`)
        // logToFile(logFile, `total prompt:\n${totalPrompt}`)
        try {
            if (completion.choices.length > 0) {
                logger.info(`final return ans:${completion.choices[0].message.content}`)
                //       logToFile(logFile, `final return ans:\n${completion.choices[0].message.content}`)
                return completion.choices[0].message.content
            }
            else
                return ""
        }
        catch (e) {
            return ""
        }
        //return completion.choices[0].message.content
    }
    public setFilePath(filePath: string) {
        this.filePath = filePath
    }
    public async GenerationType(CODE: string, typePrompt: string[] = []) {

        if (CODE.length > 2048 * 5) {
            return "too long for CODE"
        }
        let extra_prompt = "The possible types analyzed from the import information are: "
        for (const t of typePrompt) {
            extra_prompt = extra_prompt + "\n" + t
        }
        let otherPrompt = typePrompt.length > 0 ? [extra_prompt] : []
        let ans = await this.Generation(otherPrompt, CODE)
        let ansFix = ans?.replace("mask:", "")
        ansFix = ansFix?.replace("mask:", "")
        ansFix = ansFix?.replace("<mask>", "").replace("<mask>:", "")
        ansFix = ansFix?.replace("(mask)", "").replace("(mask):", "")
        if (ansFix?.startsWith(" {") || ansFix?.startsWith("{")) {
            logger.debug("initial genneration:" + ansFix)
            let newAns = await this.parseImportInfos(CODE, ansFix)
            if (newAns != null) {
                ansFix = newAns.replace("mask:", "")
            }
        }
        else {
            if (ansFix != undefined) {
                ansFix = await this.getSimiliarData(CODE, ansFix ? ansFix : "");
            }
        }
        return this.fixAns(ansFix);
    }

    private fixAns(ans: any): string {
        let replaceData = {}
        ans = ans.replace("mask:", "")
        ans = ans.replace("<mask>", "").replace("<mask>:", "")
        ans = ans.replace("(mask)", "").replace("(mask):", "")
        ans = ans.trim();
        ans = ans.replace("RefObject", "MutableRefObject")
        if (isQuotedUnionBySplit(ans)) {
            ans = "string"
        }
        // ans = ans.replace("Record<string, any>", "object")
        if (ans == "any") {
            ans = "object"
        } else if (ans.startsWith("(") && ans.endsWith(")")) {
            ans = ans.slice(1, -1)
        }
        return ans
    }
    async parseImportInfos(code: string, generation: string) {
        const project = new Project();
        const sourceFile = project.createSourceFile("temp.ts", code);
        var importInfos = sourceFile.getDescendantsOfKind(SyntaxKind.ImportDeclaration)
        let tpPackageLoader = new ThirdPackageLoader(importInfos)
        var interfaceInfo = sourceFile.getDescendantsOfKind(SyntaxKind.InterfaceDeclaration)
        let ans0: Set<any> = new Set();
        if (this.filePath != "") {
            ans0 = this.projectDataLoader.targetFileData(this.filePath, generation)
        }

        let ans: any[] = []
        if (ans0.size == 0) {
            ans = this.projectDataLoader.parseImportInfo(importInfos, generation);
        }
        else {
            ans0.forEach(item => ans.push(item))
        }
        let ans2 = tpPackageLoader.getRecomendType(generation)
        let extra_prompt = "The possible types analyzed from the import information are: "
        if (ans.length > 0 || ans2.length > 0) {
            for (const t of ans) {
                extra_prompt = extra_prompt + "\n" + t.code
            }
            for (const t of ans2) {
                extra_prompt = extra_prompt + "\n" + t
            }
            let extra_prompts = [extra_prompt];
            let newAns = await this.Generation(extra_prompts, code)
            return newAns
        }
        else {
            return generation
        }

    }

    async getSimiliarData(CODE: string, generation: string): Promise<string> {
        const project = new Project();
        const sourceFile = project.createSourceFile("temp.ts", CODE);
        var importInfos = sourceFile.getDescendantsOfKind(SyntaxKind.ImportDeclaration)
        let ans = this.projectDataLoader.findWordSimilarity(CODE, generation)
        let tpPackageLoader = new ThirdPackageLoader(importInfos)
        let ans2 = tpPackageLoader.getRecomendType(generation)
        if (ans.length > 0 || ans2.length > 0) {
            let extra_prompt = "The possible types analyzed from the import information are: "
            for (const t of ans) {
                extra_prompt = extra_prompt + "\n" + t
            }
            for (const t of ans2) {
                extra_prompt = extra_prompt + "\n" + t
            }
            let extra_prompts = [extra_prompt];
            let newAns = await this.Generation(extra_prompts, CODE)
            if (newAns != undefined) {
                return newAns
            }
            else {
                return generation
            }
        }
        return generation;
    }
    public getTotalPrompt() {
        return this.total_prompt;
    }
    private ansFix(res: string) {
        let newRes = res.replace("mask:", "")
    }
}

async function main() {
    let LLMA: LLMAgent = new LLMAgent()
}
// main()

File: Python/LLMAgent.py

from openai import OpenAI, api_key
from loguru import logger
import re
from function_methods import Function_methods
from typing import List

K = 1

def replace_union(match):
    types = [t.strip() for t in match.group(0).split('|')]
    return f"typing.Union[{', '.join(types)}]"

def convert_type_annotation(type_str: str) -> str:
    pattern = r"list\s*\[\s*([\w\.]+)\s*\]"
    match = re.search(pattern, type_str)

    if match:
        original_type = match.group(1).strip()
        return re.sub(pattern, f"list[Union[Any, {original_type}]]", type_str)
    else:
        return type_str

def wrap_with_union(s: str) -> str:

    pattern = re.compile(r'dict\[\s*([^\[\]]+?)\s*\]')

    def repl(match: re.Match) -> str:
        inner = match.group(1)
        parts = []
        for p in inner.split(','):
            t = p.strip()
            if t.lower() == 'any':
                parts.append(t)
            else:
                parts.append(f'union[any, {t}]')
        return 'dict[' + ', '.join(parts) + ']'
    return pattern.sub(repl, s)

def convert_type_annotation_tuple(type_str: str) -> str:
    pattern = r"tuple\s*\[\s*([\w\.]+)\s*\]"
    match = re.search(pattern, type_str)

    if match:
        original_type = match.group(1).strip()
        return re.sub(pattern, f"tuple[Union[Any, {original_type}]]", type_str)
    else:
        return type_str


def re_build_new_ans(ans_list:List[str]):
    ans_map = {}
    for ans in ans_list:
        if ans not in ans_map.keys():
            ans_map[ans] = 1
        else:
            ans_map[ans] += 1
    sorted_dict = dict(sorted(ans_map.items(), key=lambda item: item[1], reverse=True))
    res_list = []
    for i in sorted_dict.keys():
        res_list.append([i, (float)(sorted_dict[i])/20.0])
    return res_list


class GPT_Client:
    client: OpenAI
    right_count = 0
    total_count = 0

    @classmethod
    def __init__(cls, api_key: str):
        cls.client =OpenAI(api_key=api_key, base_url="")

    @classmethod
    def generation(cls, prompt: str, other_prompt:[str]=[]) -> str:
        base_prompt = "Next, you will be provided with a piece of Python code slicing. You will infer the variable type or function return type in Python and fill in the type annotation in <mask>. Output just only the type ,which you infer, nothing else,"
        example_prompt = """
example for your output: mask: str
The code you need to make a prediction is:\n
        """
        total_prompt = base_prompt + example_prompt
        logger.debug(other_prompt)
        for i in other_prompt:
            total_prompt = total_prompt + i+"\n"

        total_prompt = total_prompt + prompt
        cls.total_prompt = total_prompt

        logger.debug(f"total prompt:{total_prompt}")
        res = cls.client.chat.completions.create(
            model="gpt-4",
            # model="deepseek-chat",
            # model = "gpt-4o-mini",
            # model="",
            messages=[
                {"role": "user",
                 "content":
                     total_prompt
                 },
            ],
            max_tokens=30,
            temperature=0.2,
            top_p=0.3,
            frequency_penalty=0,
            presence_penalty=0,
            n=K,
            # stop=["\n", "."]
        )
        if res.choices:
            logger.debug(f"res choices length:{len(res.choices)}")
            for i in range(len(res.choices)):
                logger.info(res.choices[i].message.content)

        try:
            if res.choices:
                if len(res.choices) == 0:
                    raise Exception("No response from LLM", res)
            logger.debug(res.choices[0].message.content)
            return res.choices[0].message.content
        except:
            return ""

    @classmethod
    def Generate_Type(cls, CODE: str, typePrompt: [str] = [], other_prompt = []):
        extra_prompt = "The possible types analyzed from the import information are: "
        for i in typePrompt:
            extra_prompt = extra_prompt + i +"\n"
        if len(typePrompt)>0:
            new_prompt = [extra_prompt]
            for i in other_prompt:
                new_prompt.append(i)
            other_prompt = new_prompt
        res = cls.generation(CODE, other_prompt)
        type_rec = cls.type_recommendation(res.replace("mask:", ""), CODE)
        if len(type_rec)>0:
            for i in type_rec:
                if i not in typePrompt:
                    extra_prompt = extra_prompt + i + "\n"
            new_other_prompt = [extra_prompt]
            if len(other_prompt)>0:
                for o_p in other_prompt:
                    new_other_prompt.append(o_p)
            # other_prompt.append(extra_prompt)
            res = cls.generation(CODE, new_other_prompt)
        logger.debug(f"res:{res}")
        return cls.generation_type_fix(res)

    @classmethod
    def Generate_Type_Hint2(cls, CODE: str, typePrompt: [str] = [], other_prompt = [], K = 20):
        ans = []
        for count_times in range(K):
            extra_prompt = "The possible types analyzed from the import information are: "
            for i in typePrompt:
                extra_prompt = extra_prompt + i + "\n"
            # other_prompt = [extra_prompt].extend(other_prompt) if len(typePrompt)>0 else other_prompt

            if len(typePrompt) > 0:
                new_prompt = [extra_prompt]
                for i in other_prompt:
                    new_prompt.append(i)
            else:
                new_prompt = []
                for i in other_prompt:
                    new_prompt.append(i)
            res = cls.generation(CODE, new_prompt)
            type_rec = cls.type_recommendation(res.replace("mask:", ""), CODE)
            if len(type_rec) > 0:
                # extra_prompt = "The possible types analyzed from the import information are: "
                for i in type_rec:
                    if i not in typePrompt:
                        extra_prompt = extra_prompt + i + "\n"
                new_other_prompt = [extra_prompt]
                if len(other_prompt) > 0:
                    for o_p in other_prompt:
                        new_other_prompt.append(o_p)
                # other_prompt.append(extra_prompt)
                res = cls.generation(CODE, new_other_prompt)
            logger.debug(f"res:{res}")
            ans.append(cls.generation_type_fix(res))

        return re_build_new_ans(ans)



    @classmethod
    def type_recommendation(cls, gen:str, total_prompt:str=""):
        cls.project_analyzer = Function_methods()
        res = cls.project_analyzer.calculate_similarity_for_class(gen)
        res2 = cls.project_analyzer.calculate_similarity_for_class_name(gen)
        if len(res)>0 or len(res2)>0:
            temp_data = []
            # for i in res:
            #     logger.debug(f"recommendation type:{i[0]}, similarity:{i[1]}")
            #     temp_data.append(i[0])
            for i in res2:
                logger.debug(f"recommendation type:{i}")
                if i not in total_prompt:
                    temp_data.append(i)
            return temp_data
        return []

    @classmethod
    def get_total_prompt(cls):
        return cls.total_prompt

    @classmethod
    def generation_type_fix(cls, gen_ans:str)->str:
        new_ans = gen_ans.replace("mask:","")
        fix_dict = {"string":"str",
                    "Record": "dict",
                    "number":"int",
                    "boolean":"bool",
                    "Array":"list",
                    "Union": "typing.Union"}
        for ru in fix_dict.keys():
            new_ans = new_ans.replace(ru, fix_dict[ru])
        if new_ans.startswith("{") and new_ans.endswith("}"):
            new_ans = "dict"
        elif new_ans.lower().startswith("list") or new_ans.lower().startswith("dict"):
            new_ans = new_ans.replace("<", "[").replace(">","]")
        elif new_ans.endswith("[]"):
            new_ans = re.sub(r'(\b[\w\.]+)\[\]', r'list[\1]', new_ans)
        elif "|" in new_ans:
            new_ans = re.sub(r'\b[\w\.]+(?:\s*\|\s*[\w\.]+)+\b', replace_union, new_ans)
        new_ans = new_ans.strip()
        if new_ans.startswith("list["):
            new_ans = convert_type_annotation(new_ans)
        if new_ans.startswith("List["):
            new_ans = convert_type_annotation(new_ans.replace("List","list"))
        if new_ans.startswith("dict["):
            new_ans = wrap_with_union(new_ans)
        if new_ans.startswith("Dict["):
            new_ans = wrap_with_union(new_ans.replace("Dict", "dict"))
        if new_ans.startswith("tuple["):
            new_ans = convert_type_annotation_tuple(new_ans)
        return new_ans

if __name__ == "__main__":
    pass